### ISSUE TITLE
Non-determinism with division by zero

### ISSUE THREAD

**Issue by axic** [ewasm/design#92]
Source: https://webassembly.github.io/spec/core/_download/WebAssembly.pdf

pp48:
idiv_u𝑁 (𝑖1, 𝑖2)
• If 𝑖2 is 0, then the result is undefined.

idiv_s𝑁 (𝑖1, 𝑖2)
• If 𝑗2 is 0, then the result is undefined.
• Else if 𝑗1 divided by 𝑗2 is 2𝑁−1, then the result is undefined.

irem_u𝑁 (𝑖1, 𝑖2)
• If 𝑗2 is 0, then the result is undefined.

irem_s𝑁 (𝑖1, 𝑖2)
• If 𝑖2 is 0, then the result is undefined.

pp61:
trunc_u𝑀,𝑁 (𝑧)
(Not a problem since this is floating point only.)

pp62:
trunc_s𝑀,𝑁 (𝑧)
(Not a problem since this is floating point only.)

(raised by @holiman)

**Comment by axic** [comment]
For the case when the argument is zero could be easily solved by embedding a call to a function `check_for_zero`, which does traps on zero at run time.

**Comment by pepyakin** [comment]
Isn’t this operations are partial, i.e not defined on the full input domain space?
If so, these operators should trap instead of return non deterministic results?

**Comment by axic** [comment]
@pepyakin good catch, all these three are.

Perhaps the way forward is to specify a deterministic version of these four instructions, which are either:
a) provided by the "ewasm vm" (and the instructions replaced with a call to them) or
b) just injected (and thus copied in) by the verification process

**Comment by pepyakin** [comment]
yeah i mean that these operators are already act deterministically on all of their inputs.

So I’m not sure what you mean

**Comment by axic** [comment]
How are they deterministic with division by 0?

**Comment by pepyakin** [comment]
Sorry for terse answers, I was commuting.

1. Let's take [`idiv_un(i1, i2)`](https://webassembly.github.io/spec/core/exec/numerics.html#op-idiv-u) for example.
2. It states, that if i2 is 0, then the result is undefined.
3. In context of numeric operations this means that operator is partial, i.e if i2 is 0 the return value is empty set.
4. then, let's look at definition of the [numeric instructions](https://webassembly.github.io/spec/core/exec/instructions.html#numeric-instructions): It specifically says that 
> Where the underlying operators are partial, the corresponding instruction will trap when the result is not defined

So, in other words: I expect that division by 0 in every execution will result in trap.
Am I missing something?

**Comment by axic** [comment]
pp45:
> Some operators are partial, because they are not defined on certain inputs. Technically, an empty set of results is returned for these inputs.

pp63:
> Where the underlying operators are partial, the corresponding instruction will trap when the result is not defined. Where the underlying operators are non-deterministic, because they may return one of multiple possible NaN values, so are the corresponding instructions.

Lets see if there are appropriate test cases for this in the wasm spec.

**Comment by pepyakin** [comment]
Yeah, sure, for example take a look at [this one](https://github.com/WebAssembly/testsuite/blob/8958a44bf8b3dbf9e48dd1c44ee566e8bf682abe/i32.wast#L82)

**Comment by ehildenb** [comment]
Hmmm, seems like a bug in the spec then (that it says is "undefined" instead of saying "trap").

**Comment by ehildenb** [comment]
@axic you want to do the honors of opening an issue on the wasm spec repo?

**Comment by pepyakin** [comment]
I think that's not a bug but just unfortunate wording issue: [partial **functions**](https://en.wikipedia.org/wiki/Partial_function) are "undefined" on some of their inputs, and the instructions are always traps when functions they invoking are "undefined".

So this "undefined" has nothing to do with "undefined" from "undefined behavior".

**Comment by ehildenb** [comment]
I think the spec is pretty clear in other places where there should be a Trap, so I think it should say `Trap` instead of "undefined". I think either way we should open an issue on the spec repo to bring it to their attention and clarify our understanding.

**Comment by axic** [comment]
I think the spec could be improved, because one needs to jump around to unfold the entire story.

I did found some missing test cases though which I'll try to add and/or report.
